### PR TITLE
4.x:  Rework TakeUntil with lock-free methods

### PR DESCRIPTION
**Pull request by akarnokd**
This PR updates the internal logic of the `TakeUntil` to use a lock-free signalling algorithm for all event types.

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeUntil.cs [hunk 1]
```diff
@@ -24,128 +25,126 @@ public TakeUntil(IObservable<TSource> source, IObservable<TOther> other)
...
+            {
+                if (Interlocked.CompareExchange(ref halfSerializer, 1, 0) == 0)
                 {
-                    if (_open)
-                    {
-                        _parent.ForwardOnNext(value);
-                    }
-                    else
+                    ForwardOnNext(value);
+                    if (Interlocked.Decrement(ref halfSerializer) != 0)
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeUntil.cs]: What's that halfSerializer for ?

Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeUntil.cs]: Mutually excludes the call to `OnNext` with any calls to `OnError` or `OnCompleted`, deferring them after the `OnNext` returns.

**Comment by danielcweber** [comment]
You mind if I try to tidy this up a bit and send you a PR?

**Comment by akarnokd** [comment]
Tidy up how?

**Comment by danielcweber** [comment]
Stylistically and less noise. I'll make a try and we can discuss.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeUntil.cs [hunk 2]
```diff
@@ -24,131 +25,118 @@ public TakeUntil(IObservable<TSource> source, IObservable<TOther> other)
 
         internal sealed class _ : IdentitySink<TSource>
         {
+            private readonly OtherObserver _other;
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/TakeUntil.cs]: Field can be omitted, good to be merged otherwise.

**Pull request merged**